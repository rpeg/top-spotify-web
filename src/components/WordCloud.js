import React from 'react';
import Palette from 'google-palette';
import ReactWordcloud from 'react-wordcloud';
import { Row } from 'react-bootstrap';

import { makeSortedFrequencyArr } from '../lib/frequency';

const WordCloud = ({ genres, count }) => {
  const sortedGenres = makeSortedFrequencyArr(genres).slice(0, count);

  const words = sortedGenres.map((genre, i) => ({
    text: genre.name,
    value: genre.freq,
  }));

  const colors = Palette('tol-rainbow', count).map((c) => `#${c}`);

  const options = {
    colors,
    enableTooltip: true,
    deterministic: false,
    fontFamily: 'Karla, sans-serif',
    fontSizes: [10, 48],
    fontStyle: 'normal',
    fontWeight: 'bold',
    padding: 1,
    rotations: 2,
    rotationAngles: [0, 90],
    scale: 'sqrt',
    spiral: 'archimedean',
    transitionDuration: 1000,
  };

  return (
    count > 0 && (
    <Row>
      <div
        style={{
          height: '100%',
          width: '100%',
        }}
      >
        <ReactWordcloud options={options} words={words} />
      </div>
    </Row>
    )
  );
};

export default WordCloud;
