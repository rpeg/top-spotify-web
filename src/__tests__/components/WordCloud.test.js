import React from 'react';
import { shallow } from 'enzyme';
import ReactWordcloud from 'react-wordcloud';

import WordCloud from '../../components/WordCloud';

describe('WordCloud', () => {
  it('should not log errors in console', () => {
    const wrapper = shallow(<WordCloud genres={['a']} count={1} />);
    const spy = jest.spyOn(global.console, 'error');
    expect(wrapper).not.toBeNull();
    expect(spy).not.toHaveBeenCalled();
  });

  it('does not render without genres', () => {
    const wrapper = shallow(<WordCloud genres={[]} count={5} />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(ReactWordcloud).exists()).toBe(false);
  });

  it('does not render with count === 0', () => {
    const wrapper = shallow(<WordCloud genres={['a']} count={0} />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(ReactWordcloud).exists()).toBe(false);
  });

  it('renders with genres and count > 0', () => {
    const wrapper = shallow(<WordCloud genres={['a', 'b', 'c', 'd', 'e']} count={5} />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(ReactWordcloud).exists()).toBe(true);
  });
});
