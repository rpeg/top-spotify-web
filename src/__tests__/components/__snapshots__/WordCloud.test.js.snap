// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WordCloud does not render with count === 0 1`] = `""`;

exports[`WordCloud does not render without genres 1`] = `""`;

exports[`WordCloud renders with genres and count > 0 1`] = `
<ForwardRef
  noGutters={false}
>
  <div
    style={
      Object {
        "height": "100%",
        "width": "100%",
      }
    }
  >
    <Wordcloud
      callbacks={
        Object {
          "getWordTooltip": [Function],
        }
      }
      maxWords={100}
      minSize={
        Array [
          300,
          300,
        ]
      }
      options={
        Object {
          "colors": Array [
            "#781c81",
            "#447cbf",
            "#83ba6d",
            "#dbab3b",
            "#d92120",
          ],
          "deterministic": false,
          "enableTooltip": true,
          "fontFamily": "Karla, sans-serif",
          "fontSizes": Array [
            10,
            48,
          ],
          "fontStyle": "normal",
          "fontWeight": "bold",
          "padding": 1,
          "rotationAngles": Array [
            0,
            90,
          ],
          "rotations": 2,
          "scale": "sqrt",
          "spiral": "archimedean",
          "transitionDuration": 1000,
        }
      }
      words={
        Array [
          Object {
            "text": "e",
            "value": 1,
          },
          Object {
            "text": "d",
            "value": 1,
          },
          Object {
            "text": "c",
            "value": 1,
          },
          Object {
            "text": "b",
            "value": 1,
          },
          Object {
            "text": "a",
            "value": 1,
          },
        ]
      }
    />
  </div>
</ForwardRef>
`;
